_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +87,130 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
 }
 
+class ReferenceFrameStorage
+{
+private:
+    class FrameLocker
+    {
+    private:
+        UMC::FrameMemID m_frameId = -1;
+        mfx_UMC_FrameAllocator *m_frameAllocator = nullptr;
+        bool ValidFrameId() const
+        {
+            return (m_frameId >= 0);
+        }
+
+        void Lock()
+        {
+            if (ValidFrameId())
+            {
+                if (m_frameAllocator->IncreaseReference(m_frameId) != UMC::UMC_OK)
+                {
+                    throw vp9_exception(UMC::UMC_ERR_FAILED);
+                }
+            }
+        }
+
+        void Unlock()
+        {
+            if (ValidFrameId())
+            {
+                if (m_frameAllocator->DecreaseReference(m_frameId) != UMC::UMC_OK)
+                {
+                    throw vp9_exception(UMC::UMC_ERR_FAILED);
+                }
+            }
+        }
+
+        FrameLocker(const FrameLocker &) = delete;
+        FrameLocker(FrameLocker &&) = delete;
+        FrameLocker& operator=(const FrameLocker &) = delete;
+    public:
+        FrameLocker() = default;
+
+        FrameLocker(UMC::FrameMemID frameId, mfx_UMC_FrameAllocator *frameAllocator) :
+            m_frameId(frameId), m_frameAllocator(frameAllocator)
+        {
+            if (m_frameAllocator == nullptr)
+            {
+                throw vp9_exception(UMC::UMC_ERR_FAILED);
+            }
+            Lock();
+        }
+
+        FrameLocker& operator=(FrameLocker &&locker)
+        {
+            Unlock();
+            m_frameId = locker.m_frameId;
+            m_frameAllocator = locker.m_frameAllocator;
+            locker.m_frameId = -1;
+            return *this;
+        }
+
+        ~FrameLocker()
+        {
+            try
+            {
+                Unlock();
+            }
+            catch(...)
+            {
+                (void)MFX_STS_TRACE(MFX_ERR_UNKNOWN);
+            }
+        }
+    };
+
+    std::map<UMC::FrameMemID, FrameLocker[8]> m_lockedRefFrames;
+    std::vector<UMC::FrameMemID> m_decodedFramesIDs;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void FreeReferenceFrames(UMC::FrameMemID currFrame)
+    {
+        size_t erasedCount = m_lockedRefFrames.erase(currFrame);
+
+        if (erasedCount == 0)
+        {
+            throw vp9_exception(UMC::UMC_ERR_FAILED);
+        }
+    }
+public:
+    ReferenceFrameStorage(mfx_UMC_FrameAllocator *frameAllocator):m_frameAllocator(frameAllocator)
+    {
+    }
+
+    void LockReferenceFrames(UMC::FrameMemID currFrame, const UMC::FrameMemID ref_frame_map[UMC_VP9_DECODER::NUM_REF_FRAMES])
+    {
+        if(m_lockedRefFrames.find(currFrame) != m_lockedRefFrames.end())
+        {
+            // currFrame should not be in m_lockedRefFrames map
+            // but if it is there free resources (decrease the references).
+            FreeReferenceFrames(currFrame);
+            (void)MFX_STS_TRACE(MFX_ERR_UNKNOWN);
+        }
+
+        for (mfxI32 i = 0; i < NUM_REF_FRAMES; ++i)
```

Comment:
This line is inconsistent with declaration at line 163. I suggest to use same constant for declaration as well